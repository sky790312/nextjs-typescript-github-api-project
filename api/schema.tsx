export interface FetchGithubRepoParams {
  q: string
  sort?: string
  order?: string
  per_page?: number
  page?: number
}

export interface FetchGithubRepoRespSchema {
  incomplete_results: boolean
  total_count: number
  items: GithubRepoSchema[]
}

export interface GithubRepoSchema {
  id: number
  name: string
  html_url: string
}
