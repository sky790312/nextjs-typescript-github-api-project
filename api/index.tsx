import { FetchGithubRepoParams } from '@/api/schema'

export const baseUrl = 'https://api.github.com'

export const fetchGithubRepo = async (params: FetchGithubRepoParams) => {
  const url = new URL(`${baseUrl}/search/repositories`)
  Object.keys(params).forEach(key => url.searchParams.append(key, params[key]))

  return await fetch(String(url), {
    method: 'GET',
  }).then(response => {
    if (!response.ok) {
      throw response
    }

    return response.json()
  })
}
