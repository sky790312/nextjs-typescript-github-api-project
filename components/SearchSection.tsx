import React from 'react'
import { TextCenterContainer } from '@/GlobalStyles'

interface Props {
  handleKeyword: (keyword) => void
}

export const SearchSection: React.FC<Props> = React.memo(
  ({ handleKeyword }) => {
    const onKeywordChange = (e: React.ChangeEvent<HTMLInputElement>) =>
      handleKeyword(e.target.value)

    return (
      <TextCenterContainer>
        <h3>search github repo by keyowrd: </h3>
        <input type='text' onChange={onKeywordChange} />
      </TextCenterContainer>
    )
  },
)
