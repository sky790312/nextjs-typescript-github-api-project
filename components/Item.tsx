/* eslint-disable @typescript-eslint/camelcase */
import styled from 'styled-components'
import { GithubRepoSchema } from '@/api/schema'
import Link from 'next/link'

interface Props {
  item: GithubRepoSchema
}

export const Item: React.FC<Props> = ({ item }) => {
  return (
    <ItemContainer>
      <h3>{item.id}</h3>
      <p>
        Name: {item.name} -
        <Link href={item.html_url} passHref={true}>
          <a target='_blank' rel='noreferrer'>
            Link
          </a>
        </Link>
      </p>
    </ItemContainer>
  )
}

const ItemContainer = styled.div`
  padding: 20px;
  cursor: default;
  background-color: ${props => props.theme.colors.gray};

  &:hover {
    opacity: 0.7;
  }
`
