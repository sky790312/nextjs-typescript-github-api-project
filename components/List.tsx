import React from 'react'
import styled from 'styled-components'
import { GithubRepoSchema } from '@/api/schema'
import { Item } from '@/components/Item'

interface Props {
  items: GithubRepoSchema[]
}

export const List: React.FC<Props> = React.memo(({ items }) => {
  return (
    <ListContainer>
      {items?.map(item => (
        <Item key={item.id} item={item} />
      ))}
    </ListContainer>
  )
})

const ListContainer = styled.div`
  position: relative;
  padding: 50px;
`
