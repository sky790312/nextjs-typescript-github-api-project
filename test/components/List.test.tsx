import React from 'react'
import { render, withProviders } from '@/test/testUtils'
import { List } from '@/components/List'
import { GithubRepoSchema } from '@/api/schema'

describe('List component', () => {
  it('matches snapshot', () => {
    const items: GithubRepoSchema[] = [
      {
        id: 0,
        name: 'test',
        // eslint-disable-next-line @typescript-eslint/camelcase
        html_url: 'test url',
      },
    ]
    const { asFragment } = render(withProviders(<List items={items} />), {})
    expect(asFragment()).toMatchSnapshot()
  })
})
