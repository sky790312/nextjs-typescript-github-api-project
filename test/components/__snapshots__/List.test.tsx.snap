// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="List__ListContainer-sc-1l9sj4l-0 entigL"
  >
    <div
      class="Item__ItemContainer-wnfykx-0 kgDXOE"
    >
      <h3>
        0
      </h3>
      <p>
        Name: test -
        <a
          href="/test%20url"
          rel="noreferrer"
          target="_blank"
        >
          Link
        </a>
      </p>
    </div>
  </div>
</DocumentFragment>
`;
